### PR TITLE
Add L1 Instruction Cache

### PR DESCRIPTION
**Pull request by danbone**
Adding an L1 Instruction Cache for fetch as noted in #143 

- Updated fetch unit so that it requests a block of instructions per cycle from the instruction cache. The block is formed based on the instruction's program counters.
- Tidied up some coding style within the L2Cache/MSS.
- Added tests for new instruction cache unit.
- Changed MSS credit interfaces to be more like those used within the core.

### CODE REVIEW

**Review by klingaard — APPROVED** [review]
LGTM.  Appreciate your changing code to match the Code Style guie.

File: core/DCache.cpp [hunk 1]
```diff
@@ -99,13 +99,13 @@ namespace olympia {
         busy_ = false;
     }
 
-    void DCache::getAckFromL2Cache_(const uint32_t &ack) {
+    void DCache::getCreditsFromL2Cache_(const uint32_t &ack) {
```
Comment (arupc-vmicro) [inline core/DCache.cpp:102]: the comments below need to be updated

File: core/Fetch.hpp [hunk 2]
```diff
@@ -99,16 +117,45 @@ namespace olympia
...
+        // Buffers up instructions read from the tracefile
+        std::deque<InstPtr> ibuf_;
+
+        // Size of trace buffer (must be sized >= L1ICache bandwidth / 2B)
+        const uint32_t ibuf_capacity_;
+
+        // Fetch buffer: Holds a queue of instructions that are either
+        // waiting for an ICache hit response, or they're ready to be
+        // send to decode
+        std::deque<InstPtr> fetch_buffer_;
```
Comment (klingaard) [inline core/Fetch.hpp:135]: Consider using a `sparta::Queue` or `sparta::Buffer` here to get free pipeline collection, capacity checks, and utilization stats.

File: core/Fetch.cpp [hunk 3]
```diff
@@ -49,52 +70,156 @@ namespace olympia
...
     void Fetch::fetchInstruction_()
     {
-        const uint32_t upper = std::min(credits_inst_queue_, num_insts_to_fetch_);
+        // Prefill the ibuf with some instructions read from the tracefile
+        // keeping enough capacity to group them into cache block accesses.
+        for (uint32_t i = ibuf_.size(); i < ibuf_capacity_; ++i)
+        {
+            const auto & inst_ptr = inst_generator_->getNextInst(my_clk_);
+            if (SPARTA_EXPECT_TRUE(nullptr != inst_ptr)) {
+                ibuf_.emplace_back(inst_ptr);
```
Comment (klingaard) [inline core/Fetch.cpp:85]: Wonder if you can append to the `ibuf_` as blocks instead of individual instructions...

**Comment by arupc-vmicro** [comment]
@danbone are you planning to add a documentation describing uarch of the modeled L1 Icache and changes to the fetch unit as a part of this PR?

File: core/Fetch.cpp [hunk 4]
```diff
@@ -20,21 +22,40 @@ namespace olympia
...
+        ev_send_insts.reset(new sparta::SingleCycleUniqueEvent<>(&unit_event_set_, "send_instructions_out",
+                                                                  CREATE_SPARTA_HANDLER(Fetch, sendInstructions_)));
+
         // Schedule a single event to start reading from a trace file
         sparta::StartupEvent(node, CREATE_SPARTA_HANDLER(Fetch, initialize_));
 
+        // Capture when the simulation is stopped prematurely by the ROB i.e. hitting retire limit
+        node->getParent()->registerForNotification<bool, Fetch, &Fetch::onROBTerminate_>(
+            this, "rob_stopped_notif_channel", false /* ROB maybe not be constructed yet */);
+
```
Comment (arupc-vmicro) [inline core/Fetch.cpp:58]: Can you please clarify why this is needed?

File: core/Fetch.cpp [hunk 5]
```diff
@@ -49,52 +70,156 @@ namespace olympia
...
-        fetch_inst_event_->schedule(1);
+        ev_fetch_insts->schedule(1);
     }
 
+
     void Fetch::fetchInstruction_()
     {
-        const uint32_t upper = std::min(credits_inst_queue_, num_insts_to_fetch_);
+        // Prefill the ibuf with some instructions read from the tracefile
+        // keeping enough capacity to group them into cache block accesses.
```
Comment (arupc-vmicro) [inline core/Fetch.cpp:80]: Is there a specific reason why we need to prefill instructions from the trace? Can we not read the trace after we get the response back from Icache? When we have decoupled frontend, isn't that what we would expect to do?

**Comment by danbone** [comment]
Yeah that's fine by me, I haven't progressed the branch. Though @arupc-vmicro  did contact me to take ownership of the branch a few months back.

**Comment by jeffnye-gh** [comment]
What is the status of this PR? I see it now has conflicts, if those were fixed is this ready to go?

I have a need in BPU evals.

**Comment by arupc** [comment]
Since Knute and Jeff has reviewed this, I will unassign myself from this review. Please feel free to mege if all comments from other reviewers have been addressed.

**Comment by klingaard** [comment]
Actually, this was merged in #211.  We can safely close this PR as NOT merged.

**Pull request closed without merging**